torch/lib/c10d/reducer.cpp

```diff
@@ -1259,7 +1259,11 @@ void Reducer::finalize_backward() {
           bucket.future_work,
           "Expected bucket.future_work not to be null. "
           "This may indicate that communication hook was not properly installed.");
-      bucket.future_work->wait();
+      // Normally need to wait until the completion of communication hook,
+      // unless the communication hook skips allreduce and only does local accumulation for (post-)local SGD.
+      if (!bucket.future_work->completed()) {
+        bucket.future_work->wait();
+      }
```

Comment:
I'm not sure why we need this change for post local sgd? Just calling `wait()` should work even for post local SGD since we're retuning a completed future?

Reply:
We cannot call `wait()` again on a completed future returned by post local sgd. The training will be stuck.

Reply:
Per offline discussion, will make `ivalue::Future::wait()` here conditional and avoid extra overheads due to an additional unnecessary lock.

Reply:
Actually this change will fail some tests in `distributed_test.py`. With the help of @rohan-varma, realized that the returned type is `CUDAFuture`, which will be always a completed future and will still need to call `wait()` for synchronization.

Reverted this change.